src/objective/quantile_obj.cu

```diff
@@ -167,8 +170,10 @@ class QuantileRegression : public ObjFunction {
     common::Mean(ctx_, *base_score, &temp);
     double meanq = temp(0) * sw;
 
-    collective::Allreduce<collective::Operation::kSum>(&meanq, 1);
-    collective::Allreduce<collective::Operation::kSum>(&sw, 1);
+    if (info.IsRowSplit()) {
```

Comment:
Maybe we should make a wrapper function for `Allreduce` so that newcomers to the code base can understand these conditions?  I think we need some separation for communication logic and machine learning algorithm implementation. I'm open to any idea, at the moment it's quite difficult for a "machine learning person" to add a new algorithm and make sure it works correctly with different distributed system requirements. I can see that these conditions are necessary for metrics as well since they require labels. Is there a way to make this less intrusive by abstracting the logic under collective instead of exposing it to the machine learning part?

Reply:
The problem is `Allreduce` is a low level primitive, but column-wise split and vertical federated learning are higher level semantic changes, so they don't always map to each other. At the level of the communicator, I don't think we can determine whether we should skip a particular `Allreduce` call, even if we know what distributed/federated environment we are in.

Maybe we can come up with a clever mechanism to make this cleaner. Perhaps as a follow up?